### DISCUSSION TITLE
What are the exceptions worth retrying (the one representing transient errors) when IModel.BasicPublish fails ?

### DISCUSSION THREAD

**Discussion by EnricoMassone** [rabbitmq/rabbitmq-dotnet-client#1083] [Q&A]
Hello, 

I've already tried to ask this same question on both [stackoverflow](https://stackoverflow.com/questions/69036735/how-to-define-a-retry-policy-for-transient-errors-when-publishing-messages-to-ra) and the [Google user group](https://groups.google.com/g/rabbitmq-users/c/uPVikGan-qY). An interesting discussion has started in stackoverflow, but I still have no clear plan to proceed. 

To summarize, my scenario is the following:

- we are using the RabbitMQ .NET client from **.NET core 3.1 applications** (both ASP.NET core web applications and worker services)
- we are using the **5.1.0** version of the .NET client (migration to 6.X is in plan, but still to be started)
-  we want to improve the resiliency of our applications, so we want to add a retry policy to the code sending messages to RabbitMQ. We want to retry over transient failures only. 
- we are already using the connection auto recovery feature, but as documented [here](https://www.rabbitmq.com/dotnet-api-guide.html#publishers) attempts to call `IModel.BasicPublish` when the connection is down will result in errors and the driver leaves the handling of this scenario to the developer (re delivery of messages is **not** automatically attempted by the driver itself)

We are struggling with identifying the exceptions thrown by `IModel.BasicPublish` which represent t**ransient errors and so are worth retrying**. We know that all the exceptions thrown by the driver derive from [RabbitMQClientException](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/60226be2c6c7d1544943df6f830a6498dd40e964/projects/client/RabbitMQ.Client/src/client/exceptions/RabbitMQClientException.cs#L45) and are defined in the [`RabbitMQ.Client.Exceptions` namespace](https://github.com/rabbitmq/rabbitmq-dotnet-client/tree/v5.1.0/projects/client/RabbitMQ.Client/src/client/exceptions). 

It seems to us that, among all the client defined exceptions, only `BrokerUnreachableException`, `ConnectFailureException` and `OperationInterruptedException ` are worth retrying. All the others seem to be related with errors which are **not** transient in nature.

As an interesting sidenote the [Microsoft microservice architecture sample application](https://github.com/dotnet-architecture/eShopOnContainers/tree/31ab9b62b9fb02fb1c1eb7cadef285c5e6ca6731) has a retry policy similar to the one we would like to define. They retry on both `BrokerUnreachableException `and `SocketException`. See [here](https://github.com/dotnet-architecture/eShopOnContainers/blob/31ab9b62b9fb02fb1c1eb7cadef285c5e6ca6731/src/BuildingBlocks/EventBus/EventBusRabbitMQ/EventBusRabbitMQ.cs#L76).

Can you please provide some guidance on this topic ? 

Many thanks for helping.

**Comment by lukebakken** [comment]
https://github.com/dotnet/eShop
